src/main/java/dasniko/keycloak/authenticator/OTPAuthenticator.java

```diff
@@ -0,0 +1,106 @@
+package dasniko.keycloak.authenticator;
+
+import jakarta.ws.rs.core.Response;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.common.util.SecretGenerator;
+import org.keycloak.models.*;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import static dasniko.keycloak.authenticator.OTPConstants.*;
+
+public abstract class OTPAuthenticator implements Authenticator {
+
+	protected String getSecretCode(AuthenticatorConfigModel config) {
+		int length = Integer.parseInt(config.getConfig().get(CODE_LENGTH));
+        return SecretGenerator.getInstance().randomString(length, SecretGenerator.DIGITS);
+	}
+
+	protected Integer getTTL(AuthenticatorConfigModel config) {
+		return Integer.parseInt(config.getConfig().get(CODE_TTL));
+	}
+
+	/**
+	 * Validates that auth session context has required defined values
+	 * Will throw an error if code or ttl values are not defined
+	 * An error page will be displayed in the UI
+	 * @param  context  AuthenticationFlowContext
+	 * @return void
+	 */
+	protected void codeContextIsValid(AuthenticationFlowContext context) {
+		AuthenticationSessionModel authSession = context.getAuthenticationSession();
+		String code = authSession.getAuthNote(CODE);
+		String ttl = authSession.getAuthNote(CODE_TTL);
+		if (code == null || ttl == null) {
+			context.failureChallenge(AuthenticationFlowError.INTERNAL_ERROR,
+				context.form().createErrorPage(Response.Status.INTERNAL_SERVER_ERROR));
+		}
+	}
+
+	/**
+	 * Checks that the OTP code entered by a user matches the code in the auth session context
+	 * @param  context  AuthenticationFlowContext
+	 * @return boolean
+	 */
+	private boolean enteredCodeIsValid(AuthenticationFlowContext context) {
+		String enteredCode = context.getHttpRequest().getDecodedFormParameters().getFirst(CODE);
+		AuthenticationSessionModel authSession = context.getAuthenticationSession();
+		String code = authSession.getAuthNote(CODE);
+		return enteredCode.equals(code);
+	}
+
+	/**
+	 * Validates that the OTP code entered by a user is correct, and still valid for use in the current session.
+	 * On success, OTP flow is complete.
+	 * If either check fails, will throw an error.
+	 * An error screen will be displayed in the UI indicating the type of error.
+	 * @param  context  AuthenticationFlowContext
+	 * @param formCode String
+	 * @return void
+	 */
+	protected void validateEnteredCode(AuthenticationFlowContext context, String formCode) {
```

Comment:
Looks like this function modifies the context that is passed in as an argument. Is this standard for the `Authenticator` interface?

Reply:
Not sure i'm answering your question correctly - but i Iooked in some of the other classes that use this type of context, and it looks like similar operations are called? Like setting a user, modifying the form, etc 

https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java#L194

https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/UsernamePasswordForm.java#L51